File: codigo/10 - Model Selection & Boosting/1 - Model Selection/grid_search.py
# -*- coding: utf-8 -*-
"""
Created on Mon Jan 11 18:40:08 2021

@author: msantamaria
"""

# Grid Search

# Importar las librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el data set 
dataset = pd.read_csv("Social_Network_Ads.csv")
X = dataset.iloc[:,[2,3]].values # iloc -> index localization
y = dataset.iloc[:,4].values

# Dividir el dataset en conjunto de entrenamiento y en conjunto de testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.25,random_state = 0)
 
 # Escalado de variables
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

# Ajustar el clasificador en el Conjunto de Entrenamiento
from sklearn.svm import SVC
classifier = SVC(kernel = "rbf", random_state=0)
classifier.fit(X_train,y_train)

# Predicción de los resultados con el Conjunto de Testing
y_pred = classifier.predict(X_test)

# Elaborar una matriz de confusión
from sklearn.metrics import confusion_matrix 
cm = confusion_matrix(y_test,y_pred)

# Aplicar k-fold cross validation
from sklearn.model_selection import cross_val_score
accuracies = cross_val_score(estimator = classifier, X = X_train, y = y_train, cv = 10)
accuracies.mean()
accuracies.std()

# Aplicar la mejora de Grid Search para optimizar el modelo y sus parámetros
from sklearn.model_selection import GridSearchCV
parameters = [{"C": [1,10,100,1000], 
               "kernel": ["linear"]},
              {"C": [1,10,100,1000], 
               "kernel": ["rbf"],
               "gamma": [0.62,0.64,0.66,0.68,0.72,0.74,0.70,0.76]}]
grid_search = GridSearchCV(estimator = classifier, 
                           param_grid = parameters,
                           scoring = "accuracy",
                           cv = 10,
                           n_jobs = -1)
grid_search = grid_search.fit(X_train, y_train)
best_accuracy = grid_search.best_score_
best_parameters = grid_search.best_params_

File: codigo/10 - Model Selection & Boosting/1 - Model Selection/grid_search.R
# Grid Search

# Importar el data set
dataset = read.csv("Social_Network_Ads.csv")
dataset = dataset[,3:5]

dataset$Purchased = factor(dataset$Purchased)

# Dividir los datos en conjunto de entrenamiento y en conjunto de testing
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased,SplitRatio = 0.75)
training_set = subset(dataset,split==TRUE)
testing_set = subset(dataset,split==FALSE)

# Escalado de valores
training_set[,1:2] = scale(training_set[,1:2])
testing_set[,c(1,2)] = scale(testing_set[,c(1,2)])

# Ajustar el clasificador con el conjunto de entrenamiento.
library(e1071)
classifier = svm(formula = Purchased ~ .,
                 data = training_set,
                 type = "C-classification",
                 kernel = "radial")

# Predicción de los resultados con el conjunto de testing
y_pred = predict(classifier,newdata = testing_set[,-3])

# Crear la matriz de confusión
cm = table(testing_set[,3],y_pred)

# Aplicar algoritmo k-fold cross validation
library(caret)
folds = createFolds(training_set$Purchased, k = 10)
cv = lapply(folds, function(x){
  training_fold = training_set[-x,]
  test_fold = training_set[x,]
  classifier = svm(formula = Purchased ~ .,
                   data = training_fold,
                   type = "C-classification",
                   kernel = "radial")
  y_pred = predict(classifier,newdata = test_fold[,-3])
  cm = table(test_fold[,3],y_pred)
  accuracy = (cm[1,1]+cm[2,2])/(cm[1,1]+cm[2,2]+cm[2,1]+cm[1,2])
  return(accuracy)
})
accuracy = mean(as.numeric(cv))
accuracy_sd = sd(as.numeric(cv))

# Aplicar Grid Search para encontrar los parámetros óptimos
library(caret)
classifier = train(form = Purchased ~ .,
                   data = training_set,
                   method = "svmRadial")
classifier$bestTune
